v2/internal/coord/routing/explore.go

```diff
@@ -0,0 +1,490 @@
+	if cfg == nil {
+		cfg = DefaultExploreConfig()
+	} else if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	e := &Explore[K, N]{
+		self:     self,
+		cplFn:    cplFn,
+		rt:       rt,
+		cfg:      *cfg,
+		qryCpl:   -1,
+		schedule: schedule,
+	}
+
+	return e, nil
+}
+
+// Advance advances the state of the explore by attempting to advance its query if running.
+func (e *Explore[K, N]) Advance(ctx context.Context, ev ExploreEvent) ExploreState {
+	ctx, span := tele.StartSpan(ctx, "Explore.Advance", trace.WithAttributes(tele.AttrInEvent(ev)))
+	defer span.End()
+
+	switch tev := ev.(type) {
+	case *EventExplorePoll:
+		// ignore, nothing to do
+	case *EventExploreFindCloserResponse[K, N]:
+		return e.advanceQuery(ctx, &query.EventQueryNodeResponse[K, N]{
+			NodeID:      tev.NodeID,
+			CloserNodes: tev.CloserNodes,
+		})
+	case *EventExploreFindCloserFailure[K, N]:
+		span.RecordError(tev.Error)
+		return e.advanceQuery(ctx, &query.EventQueryNodeFailure[K, N]{
+			NodeID: tev.NodeID,
+			Error:  tev.Error,
+		})
+	default:
+		panic(fmt.Sprintf("unexpected event: %T", tev))
+	}
+
+	// if query is running, give it a chance to advance
+	if e.qry != nil {
+		return e.advanceQuery(ctx, &query.EventQueryPoll{})
+	}
+
+	// is an explore due yet?
+	cpl, ok := e.schedule.NextCpl(e.cfg.Clock.Now())
+	if !ok {
+		return &StateExploreIdle{}
+	}
+
+	// start an explore query by synthesizing a node whose key has the appropriate cpl
+	node, err := e.cplFn(e.self.Key(), cpl)
+	if err != nil {
+		return &StateExploreFailure{
+			Cpl:   cpl,
+			Error: fmt.Errorf("synthesize random node for cpl %d: %w", cpl, err),
+		}
+	}
+	seeds := e.rt.NearestNodes(node.Key(), 20)
+
+	iter := query.NewClosestNodesIter[K, N](e.self.Key())
+
+	qryCfg := query.DefaultQueryConfig()
+	qryCfg.Clock = e.cfg.Clock
+	// qryCfg.Concurrency = b.cfg.RequestConcurrency
+	// qryCfg.RequestTimeout = b.cfg.RequestTimeout
+
+	qry, err := query.NewFindCloserQuery[K, N, any](e.self, ExploreQueryID, node.Key(), iter, seeds, qryCfg)
+	if err != nil {
+		return &StateExploreFailure{
+			Cpl:   cpl,
+			Error: fmt.Errorf("start explore query for cpl %d: %w", cpl, err),
+		}
+	}
+	e.qry = qry
+	e.qryCpl = cpl
+
+	return e.advanceQuery(ctx, &query.EventQueryPoll{})
+}
+
+func (e *Explore[K, N]) advanceQuery(ctx context.Context, qev query.QueryEvent) ExploreState {
+	ctx, span := tele.StartSpan(ctx, "Explore.advanceQuery")
+	defer span.End()
+	state := e.qry.Advance(ctx, qev)
+	switch st := state.(type) {
+	case *query.StateQueryFindCloser[K, N]:
+		return &StateExploreFindCloser[K, N]{
+			Cpl:     e.qryCpl,
+			QueryID: st.QueryID,
+			Stats:   st.Stats,
+			NodeID:  st.NodeID,
+			Target:  st.Target,
+		}
+	case *query.StateQueryFinished[K, N]:
+		span.SetAttributes(attribute.String("out_state", "StateExploreFinished"))
```

Comment:
in other places I used `tele.AttrOutEvent(...)`. I also used the defer to capture the event that's passed out of this function.